bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4379,6 +4490,627 @@ void ReplicaImp::setConflictDetectionBlockId(const ClientRequestMsg &clientReqMs
+        ConcordAssert(!(req.flags() & MsgFlag::TIME_SERVICE_FLAG));
+        if ((req.flags() & MsgFlag::KEY_EXCHANGE_FLAG) || (req.flags() & MsgFlag::RECONFIG_FLAG)) {
+          numOfSpecialReqs++;
+          reqIdx++;
+          continue;
+        }
+        reqIdx++;
+      }
+      reqIter.restart();
+    }
+    executeAllPreprepareRequests(allowParallelExecution,
+                                 shouldRunRequestsInParallel,
+                                 numOfSpecialReqs,
+                                 ppMsg,
+                                 requestSet,
+                                 recoverFromErrorInRequestsExecution);
+
+  } else  // if we don't have requests in ppMsg
+  {
+    // send internal message that will call to finishExecutePrePrepareMsg
+    ConcordAssert(activeExecutions_ == 0);
+    activeExecutions_ = 1;
+    InternalMessage im = FinishExecutePrePrepareInternalMsg{ppMsg, nullptr};  // TODO(GG): check....
+    getIncomingMsgsStorage().pushInternalMsg(std::move(im));
+  }
+}
+
+void ReplicaImp::markSpecialRequests(RequestsIterator &reqIter,
+                                     char *requestBody,
+                                     uint16_t &numOfSpecialReqs,
+                                     size_t &reqIdx,
+                                     Bitmap &requestSet,
+                                     bool allowParallelExecution,
+                                     bool &shouldRunRequestsInParallel) {
+  while (reqIter.getAndGoToNext(requestBody)) {
+    ClientRequestMsg req((ClientRequestMsgHeader *)requestBody);
+    //        SCOPED_MDC_CID(req.getCid());
+    NodeIdType clientId = req.clientProxyId();
+
+    if (config_.timeServiceEnabled && (req.flags() & MsgFlag::TIME_SERVICE_FLAG) && reqIdx == 0) {
+      numOfSpecialReqs++;
+      reqIdx++;
+      continue;
+    }
+
+    // TODO(GG): should be verified in the validation of the PrePrepare . Consdier to remove this assert
+    ConcordAssert(!(req.flags() & MsgFlag::TIME_SERVICE_FLAG));
+    if ((req.flags() & MsgFlag::KEY_EXCHANGE_FLAG) || (req.flags() & MsgFlag::RECONFIG_FLAG) ||
+        (req.flags() & MsgFlag::CLIENTS_PUB_KEYS_FLAG)) {
+      numOfSpecialReqs++;
+      reqIdx++;
+      continue;
+    }
+
+    if (req.requestLength() == 0) {
+      if (clientId == currentPrimary())
+        ++numValidNoOps;  // TODO(GG): do we want empty requests from non-primary replicas?
+      reqIdx++;
+      continue;
+    }
+
+    const bool validClient = isValidClient(clientId);
+    if (!validClient) {
+      ++numInvalidClients;
+      //          LOG_WARN(GL, "The client is not valid" << KVLOG(clientId));
+      reqIdx++;
+      continue;
+    }
+
+    if (isReplyAlreadySentToClient(clientId, req.requestSeqNum())) {
+      auto replyMsg = clientsManager->allocateReplyFromSavedOne(clientId, req.requestSeqNum(), currentPrimary());
+      if (replyMsg) {
+        send(replyMsg.get(), clientId);
+      }
+      reqIdx++;
+      continue;
```

Comment:
Move all validation part to a separate function

Reply:
Kept it the same as the old code in order to compare both ways.